src/core/mesh/qgsmeshlayer.cpp

```diff
@@ -63,9 +63,9 @@ QgsMeshLayer::QgsMeshLayer( const QString &meshLayerPath,
     ok = setDataProvider( providerKey, providerOptions, flags );
   }
 
+  setLegend( QgsMapLayerLegend::defaultMeshLegend( this ) );
   if ( ok )
   {
-    setLegend( QgsMapLayerLegend::defaultMeshLegend( this ) );
```

Comment:
I am not sure, if you open existing project, why would `setDataProvider` return false? isnt the problem somewhere else (like read/write of qgis project xml?)

Reply:
when opening a project and the layer is created, the layer is not valid, it become valid later when read XML:
https://github.com/qgis/QGIS/blob/d2937653c3d7277540232e071f9a333f73f16607/src/core/project/qgsproject.cpp#L1189
and
https://github.com/qgis/QGIS/blob/d2937653c3d7277540232e071f9a333f73f16607/src/core/project/qgsproject.cpp#L1232

Reply:
It matches the other layer type's behavior, too